juniper_graphql_ws/src/lib.rs

```diff
@@ -0,0 +1,1071 @@
+                                        .to_stream(),
+                                )
+                                .boxed();
+                            s = s
+                                .chain(stream::unfold((), move |_| async move {
+                                    tokio::time::delay_for(keep_alive_interval).await;
+                                    Some((
+                                        Reaction::ServerMessage(ServerMessage::ConnectionKeepAlive),
+                                        (),
+                                    ))
+                                }))
+                                .boxed();
+                        }
+
+                        (
+                            Self::Active {
+                                config: Arc::new(config),
+                                stoppers: HashMap::new(),
+                                schema,
+                            },
+                            s,
+                        )
+                    }
+                    Err(e) => (
+                        Self::Terminated,
+                        stream::iter(vec![
+                            Reaction::ServerMessage(ServerMessage::ConnectionError {
+                                payload: ConnectionErrorPayload {
+                                    message: e.to_string(),
+                                },
+                            }),
+                            Reaction::EndStream,
+                        ])
+                        .boxed(),
+                    ),
+                },
+                _ => (Self::PreInit { init, schema }, stream::empty().boxed()),
+            },
+            Self::Active {
+                config,
+                mut stoppers,
+                schema,
+            } => {
+                let reactions = match msg {
+                    ClientMessage::Start { id, payload } => {
+                        if stoppers.contains_key(&id) {
+                            // We already have an operation with this id, so we can't start a new
+                            // one.
+                            stream::empty().boxed()
+                        } else {
+                            // Go ahead and prune canceled stoppers before adding a new one.
+                            stoppers.retain(|_, tx| !tx.is_canceled());
+
+                            if config.max_in_flight_operations > 0
+                                && stoppers.len() >= config.max_in_flight_operations
+                            {
+                                // Too many in-flight operations. Just send back a validation error.
+                                stream::iter(vec![
+                                    Reaction::ServerMessage(ServerMessage::Error {
+                                        id: id.clone(),
+                                        payload: GraphQLError::ValidationError(vec![
```

Comment:
Should we add a new error type here? I guess it is technically validation...

Reply:
🤷 I don't really feel strongly either way.